File: mobile/android/feature/kitchen/src/test/kotlin/com/us/android/feature/kitchen/kyc/GstinTest.kt
package com.us.android.feature.kitchen.kyc

import com.google.common.truth.Truth.assertThat
import com.google.common.truth.Truth.assertWithMessage
import org.junit.Test

/**
 * The Go tests' vectors (Architecture/shared/kyc/gstin_test.go), case for case.
 *
 * [refCheck] is an INDEPENDENT mod-36 used only to build vectors: it walks right
 * to left from weight 2, where production walks left to right from weight 1, so
 * a mutation in production cannot move the expected values. Every identifier
 * here is synthetic.
 */
class GstinTest {

    private fun refCheck(first14: String): Char {
        val alpha = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        require(first14.length == 14) { "refCheck needs 14 characters" }
        var weight = 2
        var sum = 0
        for (i in 13 downTo 0) {
            val cp = alpha.indexOf(first14[i])
            require(cp >= 0) { "bad character at $i" }
            val d = weight * cp
            sum += d / 36 + d % 36
            weight = if (weight == 2) 1 else 2
        }
        return alpha[(36 - sum % 36) % 36]
    }

    private fun synth(state: String, pan: String): String {
        val base = state + pan + "1Z"
        return base + refCheck(base)
    }

    private fun partOf(check: GstinCheck): GstinPart =
        (check as? GstinCheck.Invalid)?.part ?: throw AssertionError("expected Invalid, got $check")

    private fun assertNoEcho(check: GstinCheck, input: String) {
        val message = (check as? GstinCheck.Invalid)?.message ?: return
        val n = input.trim().uppercase()
        if (n.length >= 6) assertThat(message).doesNotContain(n)
        if (n.length >= 12) assertThat(message).doesNotContain(n.substring(2, 12))
    }

    // 2 9 Z Z Z P Z 0 0 0 0 Z 1 Z -> digit sums 246, 246 mod 36 = 30, check 6.
    @Test
    fun `the hand-computed golden 29ZZZPZ0000Z1Z6 validates into its parts`() {
        val golden = "29ZZZPZ0000Z1Z6"
        assertThat(refCheck(golden.take(14))).isEqualTo('6')
        assertThat(Gstin.checkDigit(golden.take(14))).isEqualTo('6')

        val valid = Gstin.check(golden) as GstinCheck.Valid
        assertThat(valid.normalized).isEqualTo(golden)
        assertThat(valid.stateCode).isEqualTo("29")
        assertThat(valid.stateName).isEqualTo("Karnataka")
        assertThat(valid.pan).isEqualTo("ZZZPZ0000Z")
        assertThat(valid.panHolderType).isEqualTo(PanHolderType.INDIVIDUAL)
        assertThat(valid.entityCode).isEqualTo('1')
        assertThat(valid.checkDigit).isEqualTo('6')
    }

    @Test
    fun `every assigned state code validates`() {
        val codes = (1..38).map { it.toString().padStart(2, '0') } + "97"
        for (code in codes) {
            val check = Gstin.check(synth(code, "ZZZCZ0000Z"))
            assertThat(check).isInstanceOf(GstinCheck.Valid::class.java)
            check as GstinCheck.Valid
            assertThat(check.stateCode).isEqualTo(code)
            assertThat(check.pan).isEqualTo("ZZZCZ0000Z")
            assertThat(check.panHolderType).isEqualTo(PanHolderType.COMPANY)
            assertThat(GstStateCodes.isAssigned(code)).isTrue()
            assertThat(GstStateCodes.nameOf(code)).isNotNull()
        }
    }

    @Test
    fun `lower case and surrounding ASCII whitespace normalise`() {
        val want = synth("27", "ZZZFZ0000Z")
        for (input in listOf(want.lowercase(), "  $want\t", " ${want.lowercase()}\n")) {
            assertThat((Gstin.check(input) as GstinCheck.Valid).normalized).isEqualTo(want)
        }
    }

    @Test
    fun `unassigned state codes fail STATE_CODE`() {
        for (code in listOf("00", "39", "40", "96", "98", "99")) {
            val input = synth(code, "ZZZPZ0000Z")
            val check = Gstin.check(input)
            assertThat(partOf(check)).isEqualTo(GstinPart.STATE_CODE)
            assertNoEcho(check, input)
            assertThat(GstStateCodes.isAssigned(code)).isFalse()
        }
    }

    @Test
    fun `every wrong check character fails CHECKSUM`() {
        val valid = synth("29", "ZZZHZ0000Z")
        val alpha = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        var wrong = 0
        for (c in alpha) {
            val input = valid.take(14) + c
            val check = Gstin.check(input)
            if (c == valid[14]) {
                assertThat(check).isInstanceOf(GstinCheck.Valid::class.java)
                continue
            }
            wrong++
            assertThat(partOf(check)).isEqualTo(GstinPart.CHECKSUM)
            assertNoEcho(check, input)
        }
        assertThat(wrong).isEqualTo(35)
    }

    @Test
    fun `layout failures fail FORMAT`() {
        val valid = synth("29", "ZZZPZ0000Z")
        val cases = mapOf(
            "fourteenth not Z" to valid.substring(0, 13) + "Y" + valid.substring(14),
            "entity code zero" to valid.substring(0, 12) + "0" + valid.substring(13),
            "inner space" to valid.substring(0, 7) + " " + valid.substring(8),
            "short" to valid.take(14),
            "long" to valid + "1",
            "empty" to "",
            "letter in state" to "2A" + valid.substring(2),
            "digit in PAN name" to valid.take(2) + "1" + valid.substring(3),
            "non-ASCII long s" to valid.take(2) + "ſ" + valid.substring(3),
            "symbol" to valid.take(14) + "#",
        )
        for ((name, input) in cases) {
            val check = Gstin.check(input)
            assertThat(check).isInstanceOf(GstinCheck.Invalid::class.java)
            assertWithMessage(name).that(partOf(check)).isEqualTo(GstinPart.FORMAT)
            assertNoEcho(check, input)
        }
    }

    @Test
    fun `an embedded PAN with no holder type fails PAN even with a valid checksum`() {
        val base = "29ABCDE1234F1Z"
        val input = base + refCheck(base)
        assertThat(input).isEqualTo("29ABCDE1234F1ZW")
        assertThat(partOf(Gstin.check(input))).isEqualTo(GstinPart.PAN)
        for (bad in listOf('D', 'E', 'K', 'Z', 'X')) {
            val b = "29ZZZ${bad}Z0000Z1Z"
            assertThat(partOf(Gstin.check(b + refCheck(b)))).isEqualTo(GstinPart.PAN)
        }
    }

    @Test
    fun `checkDigit refuses anything but fourteen upper-case alphanumerics`() {
        for (input in listOf("", "29ZZZPZ0000Z1", "29ZZZPZ0000Z1Z6", "29zzzpz0000z1z", "29ZZZPZ0000Z1#")) {
            assertThat(Gstin.checkDigit(input)).isNull()
        }
    }

    @Test
    fun `state names are exact two-character lookups`() {
        assertThat(GstStateCodes.nameOf("29")).isEqualTo("Karnataka")
        assertThat(GstStateCodes.nameOf("97")).isEqualTo("Other Territory")
        for (code in listOf("", "0", "00", "39", "99", " 29", "29 ")) {
            assertThat(GstStateCodes.nameOf(code)).isNull()
        }
    }

    @Test
    fun `the location picker offers food-service's state names without 97 and 28`() {
        val names = GstStateCodes.locationStateNames
        assertThat(names).contains("Karnataka")
        assertThat(names).contains("Andhra Pradesh")
        assertThat(names).doesNotContain("Other Territory")
        assertThat(names).doesNotContain("Andhra Pradesh (before reorganisation)")
        assertThat(names).hasSize(37)
        assertThat(names).isInOrder()
    }
}

File: Architecture/shared/kyc/gstin_test.go
package kyc

import (
	"errors"
	"fmt"
	"strings"
	"testing"
)

// refGSTINCheck is an independent mod-36 implementation used only to BUILD
// test vectors. It walks right to left starting at weight 2, the way GSTN's
// published description reads, where production walks left to right from
// weight 1. A mutation in production therefore cannot move the expected
// values. Every identifier built from it is synthetic.
func refGSTINCheck(t *testing.T, first14 string) byte {
	t.Helper()
	const alpha = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	if len(first14) != 14 {
		t.Fatalf("refGSTINCheck: need 14 characters, got %d", len(first14))
	}
	weight, sum := 2, 0
	for i := 13; i >= 0; i-- {
		cp := strings.IndexByte(alpha, first14[i])
		if cp < 0 {
			t.Fatalf("refGSTINCheck: bad character at %d", i)
		}
		d := weight * cp
		sum += d/36 + d%36
		if weight == 2 {
			weight = 1
		} else {
			weight = 2
		}
	}
	return alpha[(36-sum%36)%36]
}

// synthGSTIN builds a checksum-valid GSTIN from a state code and a synthetic
// PAN (serial 0000, which is not issued in practice).
func synthGSTIN(t *testing.T, state, pan string) string {
	t.Helper()
	base := state + pan + "1Z"
	return base + string(refGSTINCheck(t, base))
}

// assertNoEcho fails when an error's text contains the value that was
// validated (or a meaningful piece of it).
func assertNoEcho(t *testing.T, err error, input string) {
	t.Helper()
	if err == nil {
		return
	}
	msg := err.Error()
	candidates := []string{input, strings.TrimSpace(input), strings.ToUpper(strings.TrimSpace(input))}
	for _, c := range candidates {
		if len(c) >= 6 && strings.Contains(msg, c) {
			t.Errorf("error text echoes the input: %q", msg)
			return
		}
	}
	n := strings.ToUpper(strings.TrimSpace(input))
	if len(n) >= 12 && strings.Contains(msg, n[2:12]) {
		t.Errorf("error text echoes the embedded PAN: %q", msg)
	}
}

func gstinPartOf(t *testing.T, err error) GSTINPart {
	t.Helper()
	if !errors.Is(err, ErrInvalidGSTIN) {
		t.Fatalf("error %v does not match ErrInvalidGSTIN", err)
	}
	var ge *GSTINError
	if !errors.As(err, &ge) {
		t.Fatalf("error %v is not a *GSTINError", err)
	}
	return ge.Part
}

// Hand-computed golden, independent of both implementations:
// 2 9 Z Z Z P Z 0 0 0 0 Z 1 Z -> values 2 9 35 35 35 25 35 0 0 0 0 35 1 35,
// weights 1 2 1 2 ..., digit sums 2 18 35 35 35 15 35 0 0 0 0 35 1 35 = 246,
// 246 mod 36 = 30, check = 36 - 30 = 6.
func TestGSTINChecksum_HandComputedGolden(t *testing.T) {
	const golden = "29ZZZPZ0000Z1Z6"
	if got := refGSTINCheck(t, golden[:14]); got != '6' {
		t.Fatalf("reference implementation disagrees with the hand computation: %c", got)
	}
	got, err := GSTINCheckDigit(golden[:14])
	if err != nil || got != '6' {
		t.Fatalf("GSTINCheckDigit = %c, %v; want 6", got, err)
	}
	g, err := ValidateGSTIN(golden)
	if err != nil {
		t.Fatalf("ValidateGSTIN(golden): %v", err)
	}
	if g.Normalized != golden || g.StateCode != "29" || g.PAN != "ZZZPZ0000Z" ||
		g.PANHolderType != PANHolderIndividual || g.EntityCode != '1' || g.CheckDigit != '6' {
		t.Fatalf("parts = %+v", g)
	}
}

func TestValidateGSTIN_EveryAssignedStateCode(t *testing.T) {
	codes := []string{"97"}
	for i := 1; i <= 38; i++ {
		codes = append(codes, fmt.Sprintf("%02d", i))
	}
	for _, code := range codes {
		in := synthGSTIN(t, code, "ZZZCZ0000Z")
		g, err := ValidateGSTIN(in)
		if err != nil {
			t.Errorf("state %s: %v", code, err)
			continue
		}
		if g.StateCode != code || g.PAN != "ZZZCZ0000Z" || g.PANHolderType != PANHolderCompany {
			t.Errorf("state %s: parts %+v", code, g)
		}
		if _, ok := GSTStateName(code); !ok || !IsValidGSTStateCode(code) {
			t.Errorf("state %s: not in the table", code)
		}
	}
}

func TestValidateGSTIN_Normalises(t *testing.T) {
	want := synthGSTIN(t, "27", "ZZZFZ0000Z")
	for _, in := range []string{strings.ToLower(want), "  " + want + "\t", " " + strings.ToLower(want) + "\n"} {
		g, err := ValidateGSTIN(in)
		if err != nil {
			t.Errorf("ValidateGSTIN(%q): %v", in, err)
			continue
		}
		if g.Normalized != want {
			t.Errorf("Normalized = %q, want %q", g.Normalized, want)
		}
	}
}

func TestValidateGSTIN_RejectsUnassignedStateCodes(t *testing.T) {
	for _, code := range []string{"00", "39", "40", "96", "98", "99"} {
		in := synthGSTIN(t, code, "ZZZPZ0000Z")
		_, err := ValidateGSTIN(in)
		if err == nil {
			t.Errorf("state %s accepted", code)
			continue
		}
		if p := gstinPartOf(t, err); p != GSTINPartStateCode {
			t.Errorf("state %s: part %s, want STATE_CODE", code, p)
		}
		assertNoEcho(t, err, in)
		if IsValidGSTStateCode(code) {
			t.Errorf("IsValidGSTStateCode(%s) = true", code)
		}
	}
}

func TestValidateGSTIN_RejectsEveryWrongCheckCharacter(t *testing.T) {
	valid := synthGSTIN(t, "29", "ZZZHZ0000Z")
	const alpha = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	wrong := 0
	for i := 0; i < len(alpha); i++ {
		in := valid[:14] + string(alpha[i])
		_, err := ValidateGSTIN(in)
		if alpha[i] == valid[14] {
			if err != nil {
				t.Fatalf("the valid vector was refused: %v", err)
			}
			continue
		}
		wrong++
		if err == nil {
			t.Errorf("wrong check character %c accepted", alpha[i])
			continue
		}
		if p := gstinPartOf(t, err); p != GSTINPartChecksum {
			t.Errorf("check %c: part %s, want CHECKSUM", alpha[i], p)
		}
		assertNoEcho(t, err, in)
	}
	if wrong != 35 {
		t.Fatalf("tried %d wrong characters, want 35", wrong)
	}
}

func TestValidateGSTIN_FormatFailures(t *testing.T) {
	valid := synthGSTIN(t, "29", "ZZZPZ0000Z")
	cases := map[string]string{
		"fourteenth not Z":  valid[:13] + "Y" + valid[14:],
		"entity code zero":  valid[:12] + "0" + valid[13:],
		"inner space":       valid[:7] + " " + valid[8:],
		"short":             valid[:14],
		"long":              valid + "1",
		"empty":             "",
		"letter in state":   "2A" + valid[2:],
		"digit in PAN name": valid[:2] + "1" + valid[3:],
		"non-ASCII s":       valid[:2] + "ſ" + valid[3:],
		"symbol":            valid[:14] + "#",
	}
	for name, in := range cases {
		_, err := ValidateGSTIN(in)
		if err == nil {
			t.Errorf("%s: accepted", name)
			continue
		}
		if p := gstinPartOf(t, err); p != GSTINPartFormat {
			t.Errorf("%s: part %s, want FORMAT", name, p)
		}
		assertNoEcho(t, err, in)
	}
}

// 29ABCDE1234F1Z? carries a checksum-valid check character (computed by the
// reference implementation), but the embedded PAN's fourth character is D,
// which is not a holder type.
func TestValidateGSTIN_EmbeddedPANHolderType(t *testing.T) {
	base := "29ABCDE1234F1Z"
	in := base + string(refGSTINCheck(t, base))
	if in != "29ABCDE1234F1ZW" {
		t.Fatalf("reference checksum gives %s; the design recorded 29ABCDE1234F1ZW", in)
	}
	_, err := ValidateGSTIN(in)
	if err == nil {
		t.Fatal("bad holder type accepted")
	}
	if p := gstinPartOf(t, err); p != GSTINPartPAN {
		t.Fatalf("part %s, want PAN", p)
	}
	assertNoEcho(t, err, in)
	for _, bad := range []byte{'D', 'E', 'K', 'Z', 'X'} {
		b := "29ZZZ" + string(bad) + "Z0000Z1Z"
		in := b + string(refGSTINCheck(t, b))
		if _, err := ValidateGSTIN(in); err == nil || gstinPartOf(t, err) != GSTINPartPAN {
			t.Errorf("holder %c: err %v, want PAN", bad, err)
		}
	}
}

func TestGSTINCheckDigit_RejectsBadInput(t *testing.T) {
	for _, in := range []string{"", "29ZZZPZ0000Z1", "29ZZZPZ0000Z1Z6", "29zzzpz0000z1z", "29ZZZPZ0000Z1#"} {
		if _, err := GSTINCheckDigit(in); !errors.Is(err, ErrInvalidGSTIN) {
			t.Errorf("GSTINCheckDigit(%q) err = %v", in, err)
		}
	}
}

func TestGSTStateName(t *testing.T) {
	if n, ok := GSTStateName("29"); !ok || n != "Karnataka" {
		t.Errorf("29 = %q, %v", n, ok)
	}
	if n, ok := GSTStateName("97"); !ok || n != "Other Territory" {
		t.Errorf("97 = %q, %v", n, ok)
	}
	for _, c := range []string{"", "0", "00", "39", "99", " 29", "29 "} {
		if _, ok := GSTStateName(c); ok {
			t.Errorf("GSTStateName(%q) found", c)
		}
	}
}
